chia/consensus/blockchain.py

```diff
@@ -1081,49 +1081,79 @@ async def get_block_generator(
         if len(ref_list) == 0:
             return BlockGenerator(block.transactions_generator, [])
 
-        result: List[bytes] = []
-        previous_br = await self.get_block_record_from_db(block.prev_header_hash)
-        if previous_br is not None and self.height_to_hash(previous_br.height) == block.prev_header_hash:
-            # We are not in a reorg, no need to look up alternate header hashes
-            # (we can get them from height_to_hash)
-            # in the v2 database, we can look up blocks by height directly
-            # (as long as we're in the main chain)
-            result = await self.block_store.get_generators_at(block.transactions_generator_ref_list)
-        else:
-            # First tries to find the blocks in additional_blocks
-            curr = block
-            additional_height_dict = {}
-            while curr.prev_header_hash in additional_blocks:
-                prev: FullBlock = additional_blocks[curr.prev_header_hash]
-                additional_height_dict[prev.height] = prev
-                curr = prev
-
-            peak: Optional[BlockRecord] = self.get_peak()
-            prev_block_record = await self.get_block_record_from_db(curr.prev_header_hash)
-            reorg_chain: Dict[uint32, bytes32] = {}
-            if prev_block_record is not None and peak is not None:
+        generators: Dict[uint32, bytes] = {}
+
+        # The block heights in the transactions_generator_ref_list don't
+        # necessarily refer to the main chain. The generators may be found in 3
+        # different places. The additional blocks, a fork of the chain (but in
+        # the database) or in the main chain.
+
+        #              * <- block
+        # additional : |
+        #              * <- peak of fork (i.e. we have not
+        #              | :  validated blocks past this height)
+        #              | :
+        # peak -> *    | : reorg_chain
+        #          \   / :
+        #           \ /  :
+        #            *  <- fork point
+        #         :  |
+        #  main   :  |
+        #  chain  :  |
+        #         :  |
+        #         :  * <- genesis
+
+        generator_refs = set(block.transactions_generator_ref_list)
+
+        # traverse the additional blocks (if any) and resolve heights into
+        # generators
+        curr = block
+        to_remove = []
+        while curr.prev_header_hash in additional_blocks:
+            prev: FullBlock = additional_blocks[curr.prev_header_hash]
+            if prev.height in generator_refs:
+                if prev.transactions_generator is None:
+                    raise ValueError(Err.GENERATOR_REF_HAS_NO_GENERATOR)
+                generators[prev.height] = bytes(prev.transactions_generator)
+                to_remove.append(prev.height)
+            curr = prev
+        for i in to_remove:
+            generator_refs.remove(i)
```

Comment:
looks like you can remove from generator_refs inside the loop right after adding to generators i think it would be nicer